src/Domain/PasswordResetToken.php

```diff
@@ -0,0 +1,84 @@
+<?php
+
+namespace BigGive\Identity\Domain;
+
+use Doctrine\ORM\Mapping as ORM;
+use Doctrine\ORM\Mapping\ManyToOne;
+use Doctrine\ORM\Mapping\JoinColumn;
+use Exception;
+use Symfony\Component\Uid\Uuid;
+
+/**
+ * @ORM\Entity(repositoryClass="BigGive\Identity\Repository\PasswordResetTokenRepository")
+ * @ORM\Table(name="PasswordResetToken", indexes={
+ *     @ORM\Index(name="secret", columns={"secret"}),
+ * })
+ */
+class PasswordResetToken
+{
+    use TimestampsTrait;
+
+    /**
+     * @ORM\Id
+     * @ORM\Column(type="uuid", unique=true)
+     * @ORM\GeneratedValue(strategy="CUSTOM")
+     * @ORM\CustomIdGenerator(class="\Symfony\Bridge\Doctrine\IdGenerator\UuidGenerator")
+     */
+    private ?Uuid $id = null;
+
+    /**
+     * @ORM\Column(type="uuid", unique=true)
+     * @ORM\CustomIdGenerator(class="\Symfony\Bridge\Doctrine\IdGenerator\UuidGenerator")
+     */
+    private readonly Uuid $secret;
+
+    /**
+     * @ManyToOne(targetEntity="Person")
+     * @JoinColumn(name="person", referencedColumnName="id")
+     */
+    public readonly Person $person;
+
+    /**
+     * @ORM\Column(type="datetime", nullable=true)
+     */
+    private ?\DateTimeImmutable $used = null;
+
+    public function __construct(Person $person)
+    {
+        $this->person = $person;
+        $this->secret = Uuid::v4();
+        $this->createdNow();
+    }
+
+    public function toBase58Secret(): string
+    {
+        return $this->secret->toBase58();
+    }
+
+    /**
+     * @throws Exception if the token was created more than an hour before usage time, or was already used.
+     * Neither should happen in production as both conditions are checked by PasswordResetTokenRepository::findForUse
+     * but we check here also for belt-and-braces.
+     */
+    public function consume(\DateTimeImmutable $usageTime): void
+    {
+        // allow an entire minute extra in case we were really slow getting to this point after fetching from DB,
+        // or for clock-skew
+        $sixtyOneMinutes = new \DateInterval("PT61M");
+
+        if ($this->created_at < $usageTime->sub($sixtyOneMinutes)) {
+            throw new Exception('Token expired');
+        }
+
+        if ($this->isUsed()) {
+            throw new Exception('Token already used');
+        }
```

Comment:
@bdsl perhaps we shouldn’t tell a potential attacker if the token they used was _expired_ or _already used_. They can use this knowledge maliciously. I think it’s better to just return a generic decline message.

```suggestion
      if ($this->created_at < $usageTime->sub($sixtyOneMinutes) || $this->isUsed()) {
            throw new Exception('Invalid token');
       }
```

Reply:
This is throwing a generic exception so it should be uncaught and just go to our private logs, not the client machine.

Also it should never or almost never happen - as the function docblock says the DB query to select the token is only looking for valid tokens, so these are just extra checks for defence in depth, which is why we don't need to catch them.

Reply:
Good catch

Reply:
(and if it does go to our logs we'd like to know which type of error it is)